contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,503 @@
+* [Installation](#installation)
+* [Versioning](#versioning)
+* [References](#references)
+
+_Authors:_
+
+* @RenaudWasTaken - Renaud Gaubert &lt;rgaubert@NVIDIA.com&gt;
+* @jiayingz - Jiaying Zhang &lt;jiayingz@google.com&gt;
+
+# Motivation
+
+Kubernetes currently supports discovery of CPU and Memory primarily to a
+minimal extent. Very few devices are handled natively by Kubelet.
+
+It is not a sustainable solution to expect every hardware vendor to add their
+vendor specific code inside Kubernetes to make their devices usable.
+
+Instead, we want a solution for vendors to be able to advertise their resources
+to Kubelet and monitor them without writing custom Kubernetes code.
+We also want to provide a consistent and portable solution for users to
+consume hardware devices across k8s clusters.
+
+This document describes a vendor independant solution to:
+  * Discovering and representing external devices
+  * Making these devices available to the containers, using these devices,
+    scrubbing and securly sharing these devices.
+  * Health Check of these devices
+
+Because devices are vendor dependant and have their own sets of problems
+and mechanisms, the solution we describe is a plugin mechanism that may run
+in a container deployed through the DaemonSets mechanism.
+The targeted devices include GPUs, High-performance NICs, FPGAs, InfiniBand,
+Storage devices, and other similar computing resources that require vendor
+specific initialization and setup.
+
+The goal is for a user to be able to enable vendor devices (e.g: GPUs) through
+the following simple steps:
+  * `kubectl create -f http://vendor.com/device-plugin-daemonset.yaml`
+  * When launching `kubectl describe nodes`, the devices appear in the node
+    status as `extensions.kubernetes.io/vendor-device`. Note: naming
+    convention is discussed PR [#844](https://github.com/kubernetes/community/pull/844)
+
+# Use Cases
+
+ * I want to use a particular device type (GPU, InfiniBand, FPGA, etc.)
+   in my pod.
+ * I should be able to use that device without writing custom Kubernetes code.
+ * I want a consistent and portable solution to consume hardware devices
+   across k8s clusters.
+
+# Objectives
+
+1. Add support for vendor specific Devices in kubelet:
+    * Through an extension mechanism.
+    * Which allows discovery and health check of devices.
+    * Which allows hooking the runtime to make devices available in containers
+      and cleaning them up.
+2. Define a deployment mechanism for this new API.
+3. Define a versioning mechanism for this new API.
+
+# Non Objectives
+
+1. Handling heterogeneous nodes and topology related problems
+2. Collecting metrics is not part of this proposal. We will only solve
+   Health Check.
+
+# TLDR
+
+At their core, device plugins are simple gRPC servers that may run in a
+container deployed through the pod mechanism.
+
+These servers implement the gRPC interface defined later in this design
```

Comment:
Without me digging into the details of the gRPC interface (for which the existing audience is better suited), I *really* think we need to purposefully and explicitly make this and CSI ( @saad-ali @jieyu ) consistent in terms of naming and semantics as much as humanly possible.  If we play it forward we can plausibly imagine networking drivers evolving into the same pattern ( @dnardo @dcbw ), and consistency is important.

e.g. How does the kubelet (the orchestrator in general) find and enumerate available plugins?  Is it a UNIX socket?  In what directory?  Is the dir per-subsystem or omnibus?  Does it recurse (so plugins can bind mount a "private" dir) or do all plugins see each other's UNIX sockets?  What is the naming of the exposed Service?  Are there common methods across subsystems?  What is the high-level back-and-forth for operations?  Where are the responsibility boundaries?  Even common use of words like "namespaces". 

I'd like that to be an explicit goal of this effort, and for CSI to make sure it leaves room to be consistent.  Thoughts?